storage/src/main/java/com/google/cloud/trace/zipkin/StackdriverSpanConsumer.java

```diff
@@ -33,15 +38,93 @@
 
   private final TraceTranslator translator;
   private final TraceConsumer consumer;
+  private final AsyncListenableTaskExecutor executor;
 
-  public StackdriverSpanConsumer(TraceTranslator translator, TraceConsumer consumer) {
+  public StackdriverSpanConsumer(
+      TraceTranslator translator, TraceConsumer consumer, AsyncListenableTaskExecutor executor) {
     this.translator = translator;
     this.consumer = consumer;
+    this.executor = executor;
   }
 
   @Override
-  public void accept(List<Span> spans) {
-    Collection<Trace> traces = translator.translateSpans(spans);
-    consumer.receive(Traces.newBuilder().addAllTraces(traces).build());
+  public Call<Void> accept(List<Span> spans) {
+    final Traces traces =
+        Traces.newBuilder().addAllTraces(translator.translateSpans(spans)).build();
+    return new TracesCall(traces);
+  }
+
+  private class TracesCall extends Call<Void> {
+
+    private final Traces traces;
+
+    private volatile boolean canceled;
+    private AtomicBoolean executed;
+    @Nullable private volatile ListenableFuture<Void> future;
+
+    private TracesCall(Traces traces) {
+      this.traces = traces;
+      executed = new AtomicBoolean();
+    }
+
+    @Override
+    public Void execute() throws IOException {
+      markExecuted();
+      if (isCanceled()) {
+        throw new IOException("Canceled");
+      }
+      consumer.receive(traces);
+      return null;
+    }
+
+    @Override
+    public void enqueue(Callback<Void> callback) {
+      markExecuted();
+      if (isCanceled()) {
+        callback.onError(new IOException("Canceled"));
+        return;
+      }
+      try {
+        future =
```

Comment:
I wouldn't bother with making a future here as you can't really cancel it. probably simpler to just execute a runnable right?

Reply:
Cool - simplified the cancellation logic.